engine/src/cmdse.cpp

```diff
@@ -55,6 +55,466 @@ along with LiveCode.  If not see <http://www.gnu.org/licenses/>.  */
 
 #include "exec.h"
 
+template<typename T>
+struct MCEval_Auto
+{
+    ~MCEval_Auto(void) {T::drop(m_value);}
+    typename T::ValueType m_value = {};
+    
+    typename T::ValueType& operator & (void) {return m_value;}
+    typename T::ValueType operator * (void) {return m_value;}
+};
+
+
+template<typename T, MCExecValueType ExecValueType>
+struct MCEval_PrimitiveType
+{
+    typedef T ValueType;
+    typedef MCExpression *EvalType;
+    
+    static constexpr MCExecValueType exec_value_type = ExecValueType;
+    
+    static void drop(T p_value)
+    {
+    }
+    
+    template<typename Func>
+    static bool eval_async(MCExecContext& ctxt, MCExpression* p_expr, ValueType& r_value, Func &&remainder_function) __attribute__((noinline))
+    {
+        MCExecValue t_value;
+        p_expr->eval_ctxt(ctxt, t_value);
+        if (ctxt.HasError())
+        {
+            ctxt . Throw();
+            return false;
+        }
+        
+        MCExecTypeConvertAndReleaseAlways(ctxt, t_value.type, &t_value, exec_value_type, &r_value);
+        if (ctxt.HasError())
+        {
+            ctxt . Throw();
+            return false;
+        }
+        
+        return remainder_function(r_value);
+    }
+    
+    static bool eval(MCExecContext& ctxt, MCExpression* p_expr, ValueType& r_value) __attribute__((noinline))
+    {
+        MCExecValue t_value;
+        p_expr->eval_ctxt(ctxt, t_value);
+        if (ctxt.HasError())
+        {
+            return false;
+        }
+        
+        MCExecTypeConvertAndReleaseAlways(ctxt, t_value.type, &t_value, exec_value_type, &r_value);
+        if (ctxt.HasError())
+        {
+            return false;
+        }
+        
+        return true;
+    }
+};
+
+struct MCEval_BoolType: MCEval_PrimitiveType<bool, kMCExecValueTypeBool>
+{
+};
+
+struct MCEval_UIntType: MCEval_PrimitiveType<uinteger_t, kMCExecValueTypeUInt>
+{
+};
+
+struct MCEval_IntType: MCEval_PrimitiveType<integer_t, kMCExecValueTypeInt>
+{
+};
+
+struct MCEval_UInt16Type: MCEval_UIntType
+{
+    template<typename Func>
+    static bool eval_async(MCExecContext& ctxt, MCExpression* p_expr, uinteger_t& r_value, Func &&remainder_function) __attribute__((noinline))
+    {
+        uinteger_t t_value;
+        if (!MCEval_UIntType::eval_async(ctxt, p_expr, t_value, remainder_function))
```

Comment:
The thing to remember here is that you don't have the result of eval_async from UIntType until the remainder_function is called, so you need to make a remainder_function which does the value check and then passes to the remainder_function which was passed in.

Reply:
But I think this particular eval shouldn't be async